#' Linear models for microarray data (LIMMA)
#' 
#' 
#' Similarly to a t-test, Limma tests difference between means relative to the 
#' variability. But it uses Bayesian statistic to adjust variability estimation.
#' Bayesian statistics can be understood as a horse race bet: each horse doesn't 
#' have the same chance to win, if you take into account all parameters (ground, 
#' weather, driver, if a horse is tired, or not), i.e. if you take into account 
#' preliminary knowledge.
#' 
#' "Bayesian approaches offer some alternatives preferable to the ordinary (often 
#' called frequentist, because they invoke the idea of the long-term frequency 
#' of outcomes in imagined repeats of experiments or samples) methods for 
#' hypothesis testing, as well as for estimation and decision-making. Space 
#' limitations preclude a detailed review of the approach here; see Box and 
#' Tiao (1973), Berger (1985), and Carlin and Louis (1996) for longer 
#' expositions, and Schmitt (1969) for an elementary introduction. Sometimes 
#' the value of a parameter is predicted from theory, and it is more reasonable 
#' to test whether or not that value is consistent with the observed data than 
#' to calculate a confidence interval (Berger and Delampady 1987, Zellner 1987). 
#' For testing such hypotheses, what is usually desired (and what is 
#' sometimes believed to be provided by a statistical hypothesis test) is 
#' Pr[Ho | data]. What is obtained, as pointed out earlier, is 
#' P = Pr[observed or more extreme data | Ho]. Bayes' theorem offers a formula 
#' for converting between them."
#' From Johnson, D. (1999)
#' 
#' LIMMA (Smyth G. K., 2005) is the reference package to analyse microarray data. 
#' Because of evident similarity between microarray data and quantitation data, 
#' this tool can also be applied.
#' 
#' It estimates the fold changes and standard errors by fitting a linear model 
#' for the expression data.
#' Empirical Bayes and other shrinkage methods are used to borrow information 
#' across genes making the analyses stable even for experiments with small 
#' number of arrays.
#' 
#' Smyth, G. K. (2005). Limma: linear models for microarray data. 
#' Bioinformatics and Computational Biology Solutions using R and Bioconductor
#' R. Gentleman, V. Carey, S. Dudoit, R. Irizarry, W. Huber (eds.), 
#' Springer, New York, pages 397â€“420.
#' 
#' Johnson, D. (1999). 
#' The Insignificance of Statistical Significance Testing. 
#' The journal of wildlife management, 63(3), 763. 
#' doi:10.2307/3802789
#' 

# #Print-tip loess normalization:
# MA <- normalizeWithinArrays(RG)
# #Estimate the fold changes and standard errors by fitting a linear model for each gene. The
# #design matrix indicates which arrays are dye-swaps.
# fit <- lmFit(MA, design=c(-1,1,-1,1))
# #Apply empirical Bayes smoothing to the standard errors.
# fit <- eBayes(fit)

# set.seed(0)
library("limma")

#' @title Apply Limma method
#' 
#' @description Apply Limma method on non-paired data.
#' 
#' @details 
#' Apply Limma method, supposing that data has been log2-transformed before. 
#' Data should not be paired.
#'  
#' @param experiment Dataset for experiment (condition 1).
#' @param control Dataset for control (condition 2).
#' @return A data frame as generated by the topTable() function of Limma.
applyLimma <- function(experiment, control) {
    
    ## Estimate the fold changes and standard errors by fitting a linear model 
    ## for each gene. 
    ## The design matrix indicates which columns are from a different condition.
    fit <- lmFit(
        cbind(control,experiment),
        method = "ls",
        design = cbind(Grp1=1, 
                       Grp2vs1=c(rep(0,ncol(control)), 
                                 rep(1,ncol(experiment)))
        )
    )
    
    ## Apply empirical Bayes smoothing to the standard errors.
    ## Need a proportion of supposedly differentially expressed genes?
    fit <- eBayes(fit) # , proportion=0.01
    
    ## Compute fold change and p-values
    limmaVals <- topTable(fit, number=nrow(experiment), coef=2)
    
    ## Order by id
    limmaVals <- limmaVals[order(as.numeric(row.names(limmaVals))),]
    
    return(limmaVals)
}

#' @title Apply Limma method on paired data
#' 
#' @description Apply Limma method on paired data.
#' 
#' @details 
#' Apply Limma method, supposing that data has been log2-transformed before. 
#' Data should be paired, meaning that first column of experiment is paired 
#' with the first column of control, etc.
#'  
#' @param experiment Dataset for experiment (condition 1).
#' @param control Dataset for control (condition 2).
#' @return A data frame as generated by the topTable() function of Limma.
applyPairedLimma <- function(experiment, control) {
    
    Replicates <- factor(rep(c(1:ncol(control)), times=2))
    Treat <- factor(rep(c("C","T"), each=ncol(control)), levels=c("C","T"))
    
#     dupcor <- duplicateCorrelation(
#         cbind(control,experiment), 
#         block=c(1,2,3,1,2,3),
#         design=cbind(Grp1=1,Grp2vs1=c(rep(0,ncol(control)),rep(1,ncol(experiment)))))
    
    ## Estimate the fold changes and standard errors by fitting a linear model for each gene. 
    fit <- lmFit(
        cbind(control,experiment), 
        method = "ls",
        design = model.matrix(~Replicates+Treat)
    )
    
    ## Apply empirical Bayes smoothing to the standard errors.
    fit <- eBayes(fit) # , proportion=0.01
    
    ## Compute fold change and p-values
    limmaVals <- topTable(fit, number=nrow(experiment), coef="TreatT")
    
    ## Order by id
    limmaVals <- limmaVals[order(as.numeric(row.names(limmaVals))),]
    
    return(limmaVals)
}

#' @title Apply Limma and report result
#' 
#' @description Apply Limma and report result
#' 
#' @details 
#' Apply Limma between rows of two datasets. 
#' TODO Use ExpressionSet object as input
#' 
#' @param experiment Dataset for experiment (condition 1).
#' @param control Dataset for control (condition 2).
#' @param outputFolderTemp Temporary folder to store data generated at this 
#'  step of the analysis.
#' @param outputFile Report of current analysis step will be appended to 
#'  this Rmd file.
#' @param thresholdPVal Maximum threshold for the p-values (e.g 0.05)
#' @param isPaired If the samples are paired (ex: sample 1,2,3 in condition 1  
#'  vs sample 1,2,3 in condition 2). Default to FALSE
#' @return A data frame with two columns "p.values" "fold.change", in the 
#'  same order as the input data
applyAndReportLimma <- function(experiment, control, 
                                outputFolderTemp, outputFile, 
                                thresholdPVal, isPaired=FALSE) {
    if(!isPaired) {
        limmaVals <- applyLimma(experiment, control)
    } else {
        limmaVals <- applyPairedLimma(experiment, control)
    }
    
    matrixData <- data.frame(
        p.values = limmaVals$P.Value,
        fold.change = limmaVals$logFC,
        row.names = row.names(experiment)
    )
    
    reportLimma(matrixData, thresholdPVal, outputFile, outputFolderTemp)
    
    return(matrixData)
}


#' @title Report result of Limma
#' 
#' @description Report result of Limma
#' 
#' @details 
#' Report Limma on two datasets. 
#' 
#' @param matrixData A data frame with two columns "p.values" "fold.change"
#' @param thresholdPVal Maximum threshold for the p-values (e.g 0.05)
#' @param outputFile Report of current analysis step will be appended to 
#'  this Rmd file.
#' @param outputFolderTemp Temporary folder to store data generated at this 
#'  step of the analysis.
reportLimma <- function(matrixData, thresholdPVal, 
                        outputFile, outputFolderTemp) {
    cat('',
        'Significantly different proteins',
        '---------------------------------------------------------------------',
        '',
        '```{r echo=FALSE, warning=FALSE}',
        'limmaCitation <- citation("limma")',
        'limmaDescription <- packageDescription("limma")',
        '```',
        '',
        'Data analysis was performed using Limma modeling (R package version `r limmaDescription$Version`).',
        '',
        '',
        sep="\n", file=outputFile, append=TRUE)
    
    allTestsSignificanceRmd(matrixData, thresholdPVal, 
                            outputFile, outputFolderTemp)
    
    cat('',        
        '> ',
        '> Please cite these articles whenever using results from this software in a publication :',
        '> ',
        '> Method article :',
        '> Linear models and empirical Bayes methods for assessing differential expression in microarray experiments. ',
        '> Smyth, G. K. (2004). ',
        '> Statistical Applications in Genetics and Molecular Biology, Vol. 3, No. 1, Article 3.',
        '> doi:10.2202/1544-6115.1027',
        '> ',
        '> Software article (R package) :',
        '> `r limmaCitation$textVersion`',
        '',
        '---------------------------------------------------------------------',
        '',
        sep="\n", file=outputFile, append=TRUE)        
    
}
