#' Local-pooled-error
#' 
#' This method was developped by Nitin Jain, and published in Bioinformatics in 
#' 2003. It tests difference between medians relative to the local intensity 
#' variability. 
#' Its meant robustness for small sample size by error-pooling makes it 
#' interesting for the analysis of quantitative proteomics dataset, as 
#' suggested by Bantscheff et al. (2007).
#' Nevertheless, Murie et al. (2009) have pointed out in a microarray study 
#' that it may introduce a bias in false positive rate, lowering it below 
#' theoretically expected values.
#' 
#' LPE method is available from the Bioconductor repository of R packages for 
#' Bioinformatics.
#' http://www.bioconductor.org/packages/release/bioc/html/LPE.html
#' See this page for installation and usage instructions.
#' 
#' ! CAUTION !
#' LPE library is primarily used for analyzing data between two conditions. 
#' To use it for paired data, see the PLPE library. 
#' http://www.bioconductor.org/packages/2.11/bioc/html/PLPE.html
#' For using LPE under multiple conditions, use the HEM library.
#' http://www.bioconductor.org/packages/release/bioc/html/HEM.html
#' 
#' References:
#' Local-pooled-error test for identifying differentially expressed genes with 
#' a small number of replicated microarrays
#' Nitin Jain, Jayant Thatte, Thomas Braciale, Klaus Ley, Michael O'Connell, 
#' Jae K. Lee
#' Bioinformatics 2003, 19 (15): 1945-1951. 
#' doi:10.1093/bioinformatics/btg264
#' 
#' Quantitative mass spectrometry in proteomics: a critical review.
#' Marcus Bantscheff, Markus Schirle, Gavain Sweetman, Jens Rick, Bernhard Kuster
#' Analytical and Bioanalytical Chemistry 2007, 389 (4): 1017-1031
#' doi:10.1007/s00216-007-1486-6
#' 
#' Comparison of small n statistical tests of differential expression applied 
#' to microarrays
#' Carl Murie, Owen Woody, Anna Y. Lee, Robert Nadon
#' BMC Bioinformatics 2009, 10 (45)
#' doi:10.1186/1471-2105-10-45
#' 

library(LPE)


#' @title Apply LPE method
#' 
#' @description Apply LPE method
#' 
#' @details 
#' Apply Local-pooled-error method, supposing that data has been 
#' log2-transformed before. Compute p-values on the z-statistics returned by 
#' LPE.
#' This method should not be used for paired data (see method description).
#' 
#' @param experiment Dataset for experiment (condition 1).
#' @param control Dataset for control (condition 2).
#' @param binVal Size of the bins used to compute local variance. By default, 
#' 1 percent of total number.
#' @param rowNames Identifier for each row of the data set.
#' @return A data frame as generated by the lpe() function of LPE, and a 
#' column for the computed p-values.
applyLpeMethod <- function(experiment, control, binVal=0.01, rowNames) {
    
    varNaive <- baseOlig.error(control, q=binVal)
    varActivated <- baseOlig.error(experiment, q=binVal)
    
    #compute z-values using LPE method
    lpeVal <- data.frame(
        lpe(experiment, control, 
            varActivated, varNaive,
            probe.set.name=rowNames)
    )
    
    # compute p-value from z-values
    p.vals <- 2*pnorm(-abs( lpeVal[, "z.stats"] ))
    lpeVal <- cbind(lpeVal, p.vals)
    
    return(lpeVal)
}

# #' Needs an lpe.val output of LPE library
# order_by_pvalue <- function(lpe.val) {
#     order_pval <- order(lpe.val[,"p.vals"], decreasing=FALSE)
#     return(lpe.val[order_pval, ])
# }


#' @title Apply LPE method on two datasets and report result
#' 
#' @description Apply LPE method on two datasets and report result.
#' 
#' @details 
#' Apply LPE method on two datasets. 
#' TODO Use ExpressionSet object as input
#' 
#' @param experiment Dataset for experiment (condition 1).
#' @param control Dataset for control (condition 2).
#' @param outputFolderTemp Temporary folder to store data generated at this 
#'  step of the analysis.
#' @param outputFile Report of current analysis step will be appended to 
#'  this Rmd file.
#' @param thresholdPVal Maximum threshold for the p-values (e.g 0.05)
#' @return A data frame with two columns "p.values" p-values of the test, 
#'  and "fold.change" computed as the difference between the medians.
#' 
applyAndReportLPE <- function(experiment, control, 
                              outputFolderTemp, outputFile, thresholdPVal) {
    
    lpeVal <- applyLpeMethod(experiment, control, 
                             rowNames=rownames(experiment))
    
    matrixData <- data.frame(
        p.values = lpeVal[, "p.vals"],
        fold.change = lpeVal[, "median.1"] - lpeVal[, "median.2"]
    )
    
    cat('',
        'Significantly different proteins',
        '---------------------------------------------------------------------',
        '',
        '```{r citationLPE, echo=FALSE, warning=FALSE}',
        'lpeCitation <- citation("LPE")',
        'lpeDescription <- packageDescription("LPE")',
        '```',
        '',
        'Data analysis was performed using Local-pooled-error (LPE) method (R package version `r lpeDescription$Version`).',
        '',
        '',
        sep="\n", file=outputFile, append=TRUE)
    
    allTestsSignificanceRmd(matrixData, thresholdPVal, 
                            outputFile, outputFolderTemp)
    
    cat('',
        '> ',
        '> Please cite these articles whenever using results from this software in a publication :',
        '> ',
        '> Method article : ',
        '> Local-pooled-error test for identifying differentially expressed genes with a small number of replicated microarrays.',
        '> Nitin Jain, Jayant Thatte, Thomas Braciale, Klaus Ley, Michael O\'Connell, Jae K. Lee',
        '> Bioinformatics 2003, 19 (15): 1945-1951.',
        '> doi:10.1093/bioinformatics/btg264',
        '> ',
        '> Software article (R package) :',
        '> `r lpeCitation$title` by `r lpeCitation$author`',
        '> ',
        '> Usage for quantitative proteomics data suggested by :',
        '> Quantitative mass spectrometry in proteomics: a critical review.',
        '> Marcus Bantscheff, Markus Schirle, Gavain Sweetman, Jens Rick, Bernhard Kuster',
        '> Analytical and Bioanalytical Chemistry 2007, 389 (4): 1017-1031',
        '> doi:10.1007/s00216-007-1486-6',
        '> ',
        '',
        '---------------------------------------------------------------------',
        '',
        sep="\n", file=outputFile, append=TRUE)        
    
    return(matrixData)
}
